File: rclt/R/water_balance.R
# Script: water_balance.R
# Objective: Convert daily water inputs and ET into cumulative infiltration and first-passage times.
# Author: Yi Yu
# Created: 2026-06-23
# Last updated: 2026-06-23
# Inputs: Daily rainfall, irrigation, ET, and monotone cumulative infiltration.
# Outputs: Daily infiltration, cumulative infiltration, and first-passage times.
# Usage: Use exported functions after library(rclt).
# Dependencies: base R

#' Calculate daily net infiltration
#'
#' @param rain_mm,et_mm,irrigation_mm Equal-length daily water vectors.
#' @param et_factor Non-negative multiplier applied to ET.
#' @return Daily infiltration in millimetres.
#' @export
daily_infiltration <- function(
  rain_mm,
  et_mm,
  irrigation_mm = rep(0, length(rain_mm)),
  et_factor = 1
) {
  lengths <- c(length(rain_mm), length(et_mm), length(irrigation_mm))
  if (length(unique(lengths)) != 1L) {
    stop("Water-balance vectors must have equal lengths.", call. = FALSE)
  }
  values <- c(rain_mm, et_mm, irrigation_mm, et_factor)
  if (any(!is.finite(values)) ||
      any(c(rain_mm, et_mm, irrigation_mm) < 0) ||
      length(et_factor) != 1L ||
      et_factor < 0) {
    stop("Water-balance inputs must be finite and non-negative.", call. = FALSE)
  }
  pmax(rain_mm + irrigation_mm - et_factor * et_mm, 0)
}

#' Calculate cumulative net infiltration
#'
#' @inheritParams daily_infiltration
#' @return Cumulative infiltration in millimetres.
#' @export
cumulative_infiltration <- function(
  rain_mm,
  et_mm,
  irrigation_mm = rep(0, length(rain_mm)),
  et_factor = 1
) {
  cumsum(daily_infiltration(rain_mm, et_mm, irrigation_mm, et_factor))
}

#' Find first-passage times for cumulative infiltration
#'
#' @param cumulative_infiltration_mm Finite non-decreasing infiltration.
#' @param time Equal-length finite increasing time values.
#' @param target_infiltration_mm Non-negative target infiltration levels.
#' @return First time reaching each target, or `NA_real_`.
#' @export
first_passage_time <- function(
  cumulative_infiltration_mm,
  time,
  target_infiltration_mm
) {
  if (length(cumulative_infiltration_mm) != length(time)) {
    stop(
      "Cumulative infiltration and time must have equal lengths.",
      call. = FALSE
    )
  }
  if (any(!is.finite(cumulative_infiltration_mm)) ||
      any(cumulative_infiltration_mm < 0) ||
      any(diff(cumulative_infiltration_mm) < 0)) {
    stop(
      "Cumulative infiltration must be finite and non-decreasing.",
      call. = FALSE
    )
  }
  if (any(!is.finite(time)) || any(diff(time) <= 0)) {
    stop("time must be finite and strictly increasing.", call. = FALSE)
  }
  if (any(!is.finite(target_infiltration_mm)) ||
      any(target_infiltration_mm < 0)) {
    stop("Targets must be finite and non-negative.", call. = FALSE)
  }

  vapply(
    target_infiltration_mm,
    function(target) {
      index <- which(cumulative_infiltration_mm >= target)[1]
      if (is.na(index)) {
        NA_real_
      } else {
        time[index]
      }
    },
    numeric(1)
  )
}

File: python/src/cltf/water_balance.py
#!/usr/bin/env python3
"""
Script: water_balance.py
Objective: Convert daily water inputs and ET into infiltration and passage times.
Author: Yi Yu
Created: 2026-06-24
Last updated: 2026-06-24
Inputs: Daily rainfall, irrigation, ET, time, and cumulative infiltration.
Outputs: Daily/cumulative infiltration and first-passage time arrays.
Usage: Import public functions from cltf or cltf.water_balance.
Dependencies: math, numpy
"""

from __future__ import annotations

import math

import numpy as np
from numpy.typing import ArrayLike


def daily_infiltration(
    rain_mm: ArrayLike,
    et_mm: ArrayLike,
    irrigation_mm: ArrayLike | None = None,
    et_factor: float = 1.0,
) -> np.ndarray:
    """Calculate thresholded daily net infiltration in millimetres."""

    rain = np.atleast_1d(np.asarray(rain_mm, dtype=float))
    et = np.atleast_1d(np.asarray(et_mm, dtype=float))
    irrigation = (
        np.zeros_like(rain)
        if irrigation_mm is None
        else np.atleast_1d(np.asarray(irrigation_mm, dtype=float))
    )
    if len({len(rain), len(et), len(irrigation)}) != 1:
        raise ValueError("Water-balance vectors must have equal lengths")
    if (
        not np.all(np.isfinite(rain))
        or not np.all(np.isfinite(et))
        or not np.all(np.isfinite(irrigation))
        or np.any(rain < 0)
        or np.any(et < 0)
        or np.any(irrigation < 0)
        or not math.isfinite(et_factor)
        or et_factor < 0
    ):
        raise ValueError("Water-balance inputs must be finite and non-negative")

    return np.maximum(rain + irrigation - et_factor * et, 0.0)


def cumulative_infiltration(
    rain_mm: ArrayLike,
    et_mm: ArrayLike,
    irrigation_mm: ArrayLike | None = None,
    et_factor: float = 1.0,
) -> np.ndarray:
    """Calculate cumulative thresholded net infiltration in millimetres."""

    return np.cumsum(
        daily_infiltration(
            rain_mm,
            et_mm,
            irrigation_mm=irrigation_mm,
            et_factor=et_factor,
        )
    )


def first_passage_time(
    cumulative_infiltration_mm: ArrayLike,
    time: ArrayLike,
    target_infiltration_mm: ArrayLike,
) -> np.ndarray:
    """Return the first time at which each infiltration target is reached."""

    infiltration = np.atleast_1d(
        np.asarray(cumulative_infiltration_mm, dtype=float)
    )
    time_values = np.atleast_1d(np.asarray(time, dtype=float))
    targets = np.atleast_1d(
        np.asarray(target_infiltration_mm, dtype=float)
    )
    if len(infiltration) != len(time_values):
        raise ValueError(
            "Cumulative infiltration and time must have equal lengths"
        )
    if (
        not np.all(np.isfinite(infiltration))
        or np.any(infiltration < 0)
        or np.any(np.diff(infiltration) < 0)
    ):
        raise ValueError(
            "Cumulative infiltration must be finite and non-decreasing"
        )
    if (
        not np.all(np.isfinite(time_values))
        or np.any(np.diff(time_values) <= 0)
    ):
        raise ValueError("time must be finite and strictly increasing")
    if not np.all(np.isfinite(targets)) or np.any(targets < 0):
        raise ValueError("Targets must be finite and non-negative")

    indices = np.searchsorted(infiltration, targets, side="left")
    result = np.full(targets.shape, np.nan, dtype=float)
    reached = indices < len(infiltration)
    result[reached] = time_values[indices[reached]]
    return result
